packages/automerge-repo/src/DocHandle.ts

```diff
@@ -78,156 +63,90 @@ export class DocHandle<T> //
+        onUpdate: assign(({ context, event }) => {
+          const oldDoc = context.doc
+          assertEvent(event, UPDATE)
+          const { callback } = event.payload
+          const doc = callback(oldDoc)
+          return { doc }
+        }),
+        onDelete: assign(() => {
+          this.emit("delete", { handle: this })
+          return { doc: undefined }
+        }),
+        onUnavailable: () => {
+          this.emit("unavailable", { handle: this })
         },
+      },
+    }).createMachine({
+      /** @xstate-layout N4IgpgJg5mDOIC5QAoC2BDAxgCwJYDswBKAYgFUAFAEQEEAVAUQG0AGAXUVAAcB7WXAC64e+TiAAeiAOwAOAKwA6ACxSAzKqks1ATjlTdAGhABPRAFolAJksKN2y1KtKAbFLla5AX09G0WPISkVAwAMgyMrBxIILz8QiJikggAjCzOijKqLEqqybJyLizaRqYIFpbJtro5Uo7J2o5S3r4YOATECrgQADZgJADCAEoM9MzsYrGCwqLRSeoyCtra8pa5adquySXmDjY5ac7JljLJeepKzSB+bYGdPX0AYgCSAHJUkRN8UwmziM7HCgqyVcUnqcmScmcMm2ZV2yiyzkOx1OalUFx8V1aAQ63R46AgBCgJGGAEUyAwAMp0D7RSbxGagJKHFgKOSWJTJGRSCosCpKaEmRCqbQKU5yXINeTaer6LwY67YogKXH4wkkKgAeX6AH1hjQqABNGncL70xKIJQ5RY5BHOJag6wwpRyEWImQVeT1aWrVSXBXtJUqgn4Ik0ADqNCedG1L3CYY1gwA0saYqbpuaEG4pKLksKpFDgcsCjDhTnxTKpTLdH6sQGFOgAO7oKYhl5gAQNngAJwA1iRY3R40ndSNDSm6enfpm5BkWAVkvy7bpuTCKq7ndZnfVeSwuTX-HWu2AAI4AVzgQhD6q12rILxoADVIyEaAAhMLjtM-RmIE4LVSQi4nLLDIGzOCWwLKA0cgyLBoFWNy+43B0R5nheaqajqepjuMtJfgyEh-FoixqMCoKqOyhzgYKCDOq6UIeuCSxHOoSGKgop74OgABuzbdOgABGvTXlho5GrhJpxJOP4pLulT6KoMhpJY2hzsWNF0QobqMV6LG+pc+A8BAcBiP6gSfFJ36EQgKksksKxrHamwwmY7gLKB85QjBzoAWxdZdL0FnfARST8ooLC7qoTnWBU4pyC5ViVMKBQaHUDQuM4fm3EGhJBWaU7-CysEAUp3LpEpWw0WYRw2LmqzgqciIsCxWUdI2zaXlAbYdt2PZ5dJ1n5jY2iJY1ikOIcMJHCyUWHC62hRZkUVNPKta3Kh56wJ1-VWUyzhFc64JWJCtQNBBzhQW4cHwbsrVKpxPF8YJgV4ZZIWIKkiKiiNSkqZYWjzCWaQ5hFh0AcCuR3QoR74qUknBRmzholpv3OkpRQNNRpTzaKTWKbIWR5FDxm9AIkA7e9skUYCWayLILBZGoLkUSKbIyIdpxHPoyTeN4QA */
 
-        {
-          actions: {
-            /** Put the updated doc on context */
-            onUpdate: assign((context, { payload }: UpdateEvent<T>) => {
-              const { doc: oldDoc } = context
-
-              const { callback } = payload
-              const newDoc = callback(oldDoc)
-
-              return { doc: newDoc }
-            }),
-            onDelete: assign(() => {
-              this.emit("delete", { handle: this })
-              return { doc: undefined }
-            }),
-            onUnavailable: assign(context => {
-              const { doc } = context
-
-              this.emit("unavailable", { handle: this })
-              return { doc }
-            }),
+      // You can use the XState extension for VS Code to visualize this machine.
+      // Or, you can see this static visualization (last updated April 2024): https://stately.ai/registry/editor/d7af9b58-c518-44f1-9c36-92a238b04a7a?machineId=91c387e7-0f01-42c9-a21d-293e9bf95bb7
+
+      initial: "idle",
+      context: { documentId, doc },
+      on: {
+        UPDATE: { actions: "onUpdate" },
+        DELETE: ".deleted",
+      },
```

Comment:
since these events are handled the same way regarding which state we're in, we can just define them once